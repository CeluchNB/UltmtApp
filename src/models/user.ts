import { Realm } from '@realm/react'

export const DisplayUserSchema: Realm.ObjectSchema = {
    name: 'DisplayUser',
    embedded: true,
    properties: {
        _id: 'string',
        firstName: 'string',
        lastName: 'string',
        username: 'string',
        localGuest: 'bool?',
    },
}
