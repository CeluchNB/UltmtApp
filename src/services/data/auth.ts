import * as Constants from '../../utils/constants'
import { ApiError } from '../../types/services'
import { AxiosResponse } from 'axios'
import EncryptedStorage from 'react-native-encrypted-storage'
import jwt_decode from 'jwt-decode'
import { isTokenExpired, throwApiError } from '../../utils/service-utils'
import {
    login as networkLogin,
    logout as networkLogout,
    refreshToken as networkRefreshToken,
} from '../network/auth'

/**
 * Method to login
 * @param username username or email
 * @param password password
 * @returns user token
 * @throws error if backend returns an error
 */
export const login = async (
    username: string,
    password: string,
): Promise<void> => {
    try {
        const response = await networkLogin(username, password)

        const tokens = response.data.tokens
        const { access, refresh } = tokens
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
    } catch (error: any) {
        return throwApiError(error, Constants.GENERIC_LOGIN_ERROR)
    }
}

/**
 * Logout current user
 * @param token token of current user
 * @throws error if backend returns an error
 */
export const logout = async () => {
    try {
        await withToken(networkLogout)
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
    } catch (error: any) {
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
        throw throwApiError(error, Constants.GENERIC_LOGOUT_ERROR)
    }
}

/**
 * Method to refresh the user's authentication token
 * @returns the new access token
 */
export const refreshToken = async (): Promise<string> => {
    try {
        const token = await EncryptedStorage.getItem('refresh_token')
        if (!token) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }
        const response = await networkRefreshToken(token)
        const { tokens } = response.data
        const { access, refresh } = tokens
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
        return access
    } catch (error: any) {
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
        throw throwApiError(error, Constants.GENERIC_GET_TOKEN_ERROR)
    }
}

/**
 * Method to make an API call with an authentication token that could be expired.
 * Will make the refresh call and try again if the token is expired.
 * @param networkCall method to call
 * @param args arguments of method
 * @returns expected response of network call
 */
export const withToken = async (
    networkCall: (token: string, ...args: any) => Promise<AxiosResponse>,
    ...args: any
): Promise<AxiosResponse> => {
    try {
        const currentToken =
            (await EncryptedStorage.getItem('access_token')) || ''
        const response = await networkCall(currentToken, ...args)
        return response
    } catch (error: any) {
        const errorJson = error.toJSON()
        if (errorJson.status !== 401) {
            throw error
        }
        try {
            const newToken = await refreshToken()
            return await networkCall(newToken, ...args)
        } catch (error2) {
            throw throwApiError(error2, Constants.GENERIC_GET_TOKEN_ERROR)
        }
    }
}

/**
 * Determine if a user has a valid refresh token
 * @returns true if user logged in
 * @throws error if user is not currently logged in
 */
export const isLoggedIn = async (): Promise<boolean> => {
    try {
        const rToken = await EncryptedStorage.getItem('refresh_token')
        if (!rToken) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }

        const { exp: rExp } = jwt_decode(rToken) as any
        if (isTokenExpired(rExp)) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }

        // TODO: can probably delete this
        const accessToken = await EncryptedStorage.getItem('access_token')
        if (!accessToken) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }
        return true
    } catch (error) {
        throw throwApiError(error, Constants.GENERIC_GET_TOKEN_ERROR)
    }
}

/**
 * Method to determine if a user has a valid access token
 * @returns boolean if user has valid access token
 */
export const hasValidAccessToken = async (): Promise<boolean> => {
    const token = await EncryptedStorage.getItem('access_token')
    if (!token) {
        return false
    }

    const { exp } = jwt_decode(token) as any
    return !isTokenExpired(exp)
}

/**
 * Method to refresh the user's token if necessary. Used for live actions
 * since sockets give no response to determine if a reauth is necessary.
 */
export const refreshTokenIfNecessary = async () => {
    const hasToken = await hasValidAccessToken()
    if (!hasToken) {
        await refreshToken()
    }
}
