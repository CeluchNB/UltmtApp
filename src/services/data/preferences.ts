import EncryptedStorage from 'react-native-encrypted-storage'

export const setDarkMode = async (isDark: boolean) => {
    await EncryptedStorage.setItem('dark_mode', isDark.toString())
}

export const isDarkMode = async (): Promise<boolean> => {
    return (await EncryptedStorage.getItem('dark_mode')) !== 'false'
}
