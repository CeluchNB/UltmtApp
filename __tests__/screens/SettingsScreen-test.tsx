import * as AuthData from '../../src/services/data/auth'
import * as Preferences from '../../src/services/data/preferences'
import * as UserData from '../../src/services/data/user'
import { NavigationContainer } from '@react-navigation/native'
import { Provider } from 'react-redux'
import React from 'react'
import SettingsScreen from '../../src/screens/SettingsScreen'
import { fetchProfileData } from '../../fixtures/data'
import { setProfile } from '../../src/store/reducers/features/account/accountReducer'
import store from '../../src/store/store'
import {
    SecureEditField,
    SettingsScreenProps,
} from '../../src/types/navigation'
import { act, fireEvent, render } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const navigate = jest.fn()
const props: SettingsScreenProps = {
    navigation: { navigate } as any,
    route: { params: {} } as any,
}

beforeEach(() => {
    store.dispatch(setProfile(fetchProfileData))
    jest.resetAllMocks()
})

it('should match snapshot', () => {
    const snapshot = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    ).toJSON()

    expect(snapshot).toMatchSnapshot()
})

it('should handle logout press', async () => {
    const spy = jest
        .spyOn(AuthData, 'logout')
        .mockReturnValue(Promise.resolve())

    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('Sign Out')

    fireEvent.press(button)
    expect(spy).toHaveBeenCalled()
    // should not need this
    await act(async () => {})
    expect(navigate).toHaveBeenCalledWith('Tabs', {
        params: { screen: 'Login' },
        screen: 'Account',
    })
})

it('should handle private account switch press', async () => {
    const spy = jest.spyOn(UserData, 'setPrivate')

    const { getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const switchEl = getByTestId('private-switch')

    await act(async () => {
        fireEvent(switchEl, 'valueChange')
    })

    expect(spy).toHaveBeenCalledWith(true)
})

it('should handle first name edit', async () => {
    const spy = jest
        .spyOn(UserData, 'changeName')
        .mockReturnValueOnce(
            Promise.resolve({ ...fetchProfileData, firstName: 'newfirst' }),
        )

    const { getAllByTestId, getByPlaceholderText, getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const editButton = getAllByTestId('ef-edit-button')[0]
    fireEvent.press(editButton)
    await act(async () => {})

    const editField = getByPlaceholderText('first')
    fireEvent.changeText(editField, 'newfirst')
    await act(async () => {})

    const submitButton = getAllByTestId('ef-submit-button')[0]
    fireEvent.press(submitButton)
    await act(async () => {})

    expect(spy).toHaveBeenCalledWith('newfirst', 'last')
    const newFirstText = getByText('newfirst')
    expect(newFirstText).toBeTruthy()
})

it('should handle first name edit error', async () => {
    const spy = jest
        .spyOn(UserData, 'changeName')
        .mockReturnValueOnce(Promise.reject({ message: 'Error message' }))

    const { getAllByTestId, getByPlaceholderText, getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const editButton = getAllByTestId('ef-edit-button')[0]
    fireEvent.press(editButton)
    await act(async () => {})

    const editField = getByPlaceholderText('first')
    fireEvent.changeText(editField, 'newfirst')
    await act(async () => {})

    const submitButton = getAllByTestId('ef-submit-button')[0]
    fireEvent.press(submitButton)
    await act(async () => {})

    expect(spy).toHaveBeenCalledWith('newfirst', 'last')
    const errorMessage = getByText('Error message')
    expect(errorMessage).toBeTruthy()
})

it('should handle last name edit', async () => {
    const spy = jest
        .spyOn(UserData, 'changeName')
        .mockReturnValueOnce(
            Promise.resolve({ ...fetchProfileData, lastName: 'newlast' }),
        )

    const { getAllByTestId, getByPlaceholderText, getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const editButton = getAllByTestId('ef-edit-button')[1]
    fireEvent.press(editButton)
    await act(async () => {})

    const editField = getByPlaceholderText('last')
    fireEvent.changeText(editField, 'newlast')
    await act(async () => {})

    const submitButton = getAllByTestId('ef-submit-button')[0]
    fireEvent.press(submitButton)
    await act(async () => {})

    expect(spy).toHaveBeenCalledWith('first', 'newlast')
    const newLastText = getByText('newlast')
    expect(newLastText).toBeTruthy()
})

it('should handle last name edit error', async () => {
    const spy = jest
        .spyOn(UserData, 'changeName')
        .mockReturnValueOnce(Promise.reject({ message: 'Error message' }))

    const { getAllByTestId, getByPlaceholderText, getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const editButton = getAllByTestId('ef-edit-button')[1]
    fireEvent.press(editButton)
    await act(async () => {})

    const editField = getByPlaceholderText('last')
    fireEvent.changeText(editField, 'newlast')
    await act(async () => {})

    const submitButton = getAllByTestId('ef-submit-button')[0]
    fireEvent.press(submitButton)
    await act(async () => {})

    expect(spy).toHaveBeenCalledWith('first', 'newlast')
    const errorMessage = getByText('Error message')
    expect(errorMessage).toBeTruthy()
})

it('should navigate to edit email', async () => {
    const { getAllByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const editButton = getAllByTestId('ef-edit-button')[2]
    fireEvent.press(editButton)
    await act(async () => {})

    expect(navigate).toHaveBeenCalledWith('SecureEdit', {
        title: 'email',
        value: fetchProfileData.email,
        field: SecureEditField.EMAIL,
    })
})

it('should navigate to edit password', async () => {
    const { getAllByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const editButton = getAllByTestId('ef-edit-button')[3]
    fireEvent.press(editButton)
    await act(async () => {})

    expect(navigate).toHaveBeenCalledWith('SecureEdit', {
        title: 'password',
        value: 'New password',
        field: SecureEditField.PASSWORD,
    })
})

it('should delete account', async () => {
    const spy = jest
        .spyOn(UserData, 'deleteAccount')
        .mockReturnValueOnce(Promise.resolve())

    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('Delete Account')
    fireEvent.press(button)
    await act(async () => {})

    const yesButton = getByText('Yes')
    fireEvent.press(yesButton)
    await act(async () => {})

    expect(spy).toHaveBeenCalled()
    expect(navigate).toHaveBeenCalledWith('Tabs', {
        params: { screen: 'Login' },
        screen: 'Account',
    })
})

it('should handle delete account error', async () => {
    const spy = jest
        .spyOn(UserData, 'deleteAccount')
        .mockReturnValueOnce(Promise.reject({ message: 'Error message' }))

    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('Delete Account')
    fireEvent.press(button)
    await act(async () => {})

    const yesButton = getByText('Yes')
    fireEvent.press(yesButton)
    await act(async () => {})

    expect(spy).toHaveBeenCalled()
    const errorMessage = getByText('Error message')
    expect(errorMessage).toBeTruthy()
    expect(navigate).not.toHaveBeenCalled()
})

it('should close delete account modal', async () => {
    const spy = jest
        .spyOn(UserData, 'deleteAccount')
        .mockReturnValueOnce(Promise.resolve())

    const { getByText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('Delete Account')
    fireEvent.press(button)
    await act(async () => {})

    const noButton = getByText('No')
    fireEvent.press(noButton)
    await act(async () => {})

    const yesButton = await queryByText('Yes')
    expect(yesButton).toBe(null)

    expect(spy).not.toHaveBeenCalled()
    expect(navigate).not.toHaveBeenCalledWith('Login')
})

// Not 100% sure why, but this test needs to be at the end for the
// suite to function properly, should consider investigating this
// in the future.
it('should handle dark mode switch press', async () => {
    const spy = jest.spyOn(Preferences, 'setDarkMode')
    jest.spyOn(Preferences, 'isDarkMode').mockReturnValue(
        Promise.resolve(false),
    )

    const { getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <SettingsScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const switchEl = getByTestId('dark-mode-switch')

    await act(async () => {
        fireEvent(switchEl, 'valueChange')
    })

    expect(spy).toHaveBeenCalledWith(false)

    const screen = getByTestId('screen')
    expect(screen.props.style.backgroundColor).toBe('#ffffff')
})
