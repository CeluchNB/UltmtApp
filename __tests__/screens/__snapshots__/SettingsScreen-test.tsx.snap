// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#303030",
      "height": "100%",
      "opacity": 1,
    }
  }
  testID="screen"
>
  <RCTScrollView>
    <View>
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#ffd600",
              "fontSize": 61.64383561643835,
              "fontWeight": "bold",
            },
            Object {
              "alignSelf": "center",
              "flex": 1,
              "margin": 10,
            },
          ]
        }
      >
        Settings
      </Text>
      <View
        style={
          Object {
            "alignSelf": "center",
            "width": "75%",
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#ff0000",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Sign Out
              </Text>
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#ffd600",
              "flex": 1,
              "fontSize": 61.64383561643835,
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginTop": 10,
            }
          }
        >
          Device
        </Text>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 41.0958904109589,
                "fontWeight": "bold",
              }
            }
          >
            Dark Mode
          </Text>
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#31adff"
            style={
              Object {
                "height": 31,
                "width": 51,
              }
            }
            testID="dark-mode-switch"
            thumbTintColor="#ffd600"
            tintColor="#ababab"
            value={true}
          />
        </View>
        <Text
          style={
            Object {
              "color": "#ffd600",
              "flex": 1,
              "fontSize": 61.64383561643835,
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginTop": 10,
            }
          }
        >
          Account
        </Text>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 41.0958904109589,
                "fontWeight": "bold",
              }
            }
          >
            Public
          </Text>
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#31adff"
            style={
              Object {
                "height": 31,
                "width": 51,
              }
            }
            testID="private-switch"
            thumbTintColor="#ffd600"
            tintColor="#ababab"
            value={true}
          />
        </View>
        <View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffd600",
                  "fontSize": 41.0958904109589,
                  "fontWeight": "bold",
                  "textAlignVertical": "center",
                }
              }
            >
              First Name
            </Text>
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="ef-edit-button"
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#ffd600",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "lineHeight": 24,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Û∞è´
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#ababab",
                    "fontSize": 41.0958904109589,
                  }
                }
              >
                first
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffd600",
                  "fontSize": 41.0958904109589,
                  "fontWeight": "bold",
                  "textAlignVertical": "center",
                }
              }
            >
              Last Name
            </Text>
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="ef-edit-button"
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#ffd600",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "lineHeight": 24,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Û∞è´
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#ababab",
                    "fontSize": 41.0958904109589,
                  }
                }
              >
                last
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffd600",
                  "fontSize": 41.0958904109589,
                  "fontWeight": "bold",
                  "textAlignVertical": "center",
                }
              }
            >
              Email
            </Text>
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="ef-edit-button"
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#ffd600",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "lineHeight": 24,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Û∞è´
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#ababab",
                    "fontSize": 41.0958904109589,
                  }
                }
              >
                test@email.com
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffd600",
                  "fontSize": 41.0958904109589,
                  "fontWeight": "bold",
                  "textAlignVertical": "center",
                }
              }
            >
              Password
            </Text>
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="ef-edit-button"
            >
              <View>
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#ffd600",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "lineHeight": 24,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Û∞è´
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#ababab",
                    "fontSize": 41.0958904109589,
                  }
                }
              >
                *****
              </Text>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#ff0000",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Sign Out All Devices
              </Text>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "#ff0000",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 2,
              "marginTop": 10,
              "minWidth": 64,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 0.75,
                "width": 0,
              },
              "shadowOpacity": 0.24,
              "shadowRadius": 1.5,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Delete Account
              </Text>
            </View>
          </View>
        </View>
        <Modal
          animationType="slide"
          hardwareAccelerated={false}
          onRequestClose={[Function]}
          transparent={true}
          visible={false}
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
